### PR TITLE
Log uncaught exception

### PR DESCRIPTION
**Pull request by dmitri-d**
Currently when the terminate handler handles an uncaught exception, it provides limited information about what happened and what exception was thrown. This change attempts to log additional details about the exception. 

Commit Message: Log uncaught exceptions when possible
Additional Description: Provide more details when possible when an uncaught exception is handled in terminate handler.
Risk Level: low

### CODE REVIEW

File: source/exe/terminate_handler.cc
Comment (jmarantz) [inline source/exe/terminate_handler.cc:15]: s/currentException/current_exception/

Also I'd use an explicit type here ([std::exception_ptr](https://en.cppreference.com/w/cpp/error/exception_ptr)) per style guide; it's not obvious from call-site.

File: source/exe/terminate_handler.cc
Comment (jmarantz) [inline source/exe/terminate_handler.cc:16]: != nullptr

File: source/exe/terminate_handler.cc
Comment (jmarantz) [inline source/exe/terminate_handler.cc:22]: can't you capture the std::exception as a fallback here and also call e.what() for more detail?

File: source/exe/terminate_handler.cc
Comment (jmarantz) [inline source/exe/terminate_handler.cc:20]: nit: as much as I prefer using absl for this: you can just use the built-in fmtlib support in ENVOY_LOG in this case:
```
ENVOY_LOG(critical, "std::terminate called! Uncaught EnvoyException {}, see trace.", e.what());
```

**Review by jmarantz — COMMENTED** [review]
This is awesome; just a few minor comments.

Is there a way to add tests for this? One way to make this easier is to factor out the log-generating part of this into an exposed helper and unit-test that, so the abort() is not called.

File: source/exe/terminate_handler.cc
Comment (jmarantz) [reply source/exe/terminate_handler.cc:22]: oh...maybe you have to catch std::exception in a block first so you have something to call "what()" on. In what scenarios would the "..." clause be reached if you catch std::exception first?

Comment (dmitri-d) [reply source/exe/terminate_handler.cc:22]: > In what scenarios would the "..." clause be reached if you catch std::exception first?

When the thrown exception isn't derived from std::exception, which is unlikely but possible.

**Comment by dmitri-d** [comment]
Apologies for rebasing my commits, I needed to add signatures and signoffs to pacify the CI.

File: source/exe/terminate_handler.cc
Comment (dmitri-d) [reply source/exe/terminate_handler.cc]: done

File: source/exe/terminate_handler.cc
Comment (dmitri-d) [reply source/exe/terminate_handler.cc]: Added logging of exceptions derived from std::exception

File: source/exe/terminate_handler.cc
Comment (dmitri-d) [reply source/exe/terminate_handler.cc]: done

File: source/exe/terminate_handler.cc
Comment (dmitri-d) [reply source/exe/terminate_handler.cc]: done

**Comment by dmitri-d** [comment]
> Is there a way to add tests for this? One way to make this easier is to factor out the log-generating part of this into an exposed helper and unit-test that, so the abort() is not called.

I don't think there's a way to substitute the default logger with a custom, an in-memory one? If so, this is tricky to test, as the only state we change here is a side-effect in stderr. Reading from stderr when running concurrently with other tests (which also may output to stderr) is going to be messy.

**Comment by jmarantz** [comment]
There's infrastructure in Envoy to test log messages: https://github.com/envoyproxy/envoy/blob/83d633d9ff0f0a9ea550ffda2fef20f7f9f3ede1/test/test_common/logging.h#L163

**Comment by dmitri-d** [comment]
ping @jmarantz

File: source/exe/terminate_handler.cc
Comment (jmarantz) [inline source/exe/terminate_handler.cc:22]: nit: should current_exception be a const ref?

File: test/exe/terminate_handler_test.cc
Comment (jmarantz) [inline test/exe/terminate_handler_test.cc:36]: nit: put the ifdef around the message rather than around the whole test.

**Review by jmarantz — COMMENTED** [review]
lgtm modulo two small nits. Can you fix those and I'll assign a senior maintainer pass now?

/assign-from @envoyproxy/senior-maintainers

File: tools/code_format/config.yaml
Comment (jmarantz) [inline tools/code_format/config.yaml:249]: @tyxia FYI -- thanks for being on top of thread-catching issues.

I think it does make sense to exclude terminate_handler.cc from the lint check because it will re-throw. I do expect this to run on the data plane, which is where we don't generally catch exceptions.

One thought I just had, though, @dmitri-d -- we really don't want to establish a pattern of circumventing the lint check when it's inconvenient. But I think in this case what's really happening is a re-throw after the catch, which is OK.

WDYT of instead of disabling the 'catch' check by listing the file here, we instead put a RETHROW macro in source/common/common/thread.h ? That would be cleaner IMO and avoid reinforcing a pattern of circumventing checks.

File: source/exe/terminate_handler.cc
Comment (dmitri-d) [reply source/exe/terminate_handler.cc:22]: What would we try to achieve with that? It's a pointer-like object, on most platforms of the same size as a pointer.  The convention appears to be to pass it by value.

Comment (ggreenway) [reply source/exe/terminate_handler.cc:22]: I think it's ok to leave as is, given that it's a pointer already

File: tools/code_format/config.yaml
Comment (ggreenway) [reply tools/code_format/config.yaml:249]: I feel like this is a very special case: it's just trying to log some extra information before dying, due to an unknown bug (I think it's safe to say it's always a bug to have an uncaught exception in envoy). I'd be ok just leaving this as-is. Nobody would ever look at this code and think it's a pattern to copy/paste elsewhere in the code.

Comment (dmitri-d) [reply tools/code_format/config.yaml:249]: I disagree -- it's not a circumvention of the check, but rather a deliberate skip. My understanding is that while the exact behaviour is unspecified, all implementations will end up calling the handler on the same thread where the exception was thrown, which means we can end up re-throwing an exception on a worker thread -- the exact thing TRY macros are guarding against.

**Review by ggreenway — APPROVED** [review]

File: source/exe/terminate_handler.cc
Comment (jmarantz) [reply source/exe/terminate_handler.cc:22]: got i; I saw "_ptr" and was thinking about Envoy's style of using that as a nickname for unique_ptr.

**Comment by dmitri-d** [comment]
Thanks for the reviews, @jmarantz and @ggreenway.

**Pull request merged**
